Update native-modules-async.md to reflect `Task<T>` support in 0.63 (#208)
microsoft/react-native-windows-samples: docs/native-modules-async.md

### SimpleHttpModule in C#

The native module support for C# supports the common async patterns established in C# using `async`, `await` and `Task<T>`.

> To expose the module to javascript you need to declare a C# class. To indicate it should be exposed to javascript, you annotate with an [attribute](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/attributes) like:

Reviewer: ```suggestion
To expose the module to JavaScript you need to declare a C# class. To indicate it should be exposed to JavaScript, you annotate with an [attribute](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/attributes) like:
```

```c#
namespace NativeModuleSample
{
  [ReactModule]
  class SimpleHttpModule
  {
    // Methods go here.
  }
}
```

> This makes an object available to JavaScript via the expression `NativeModules.SimpleHttpModule`. By default it will use the name of the class as the name in JavaScript. If you don't want the name of the class to match the name in JavaScript, i.e. you want to access the module the expression `NativeModules.CustomModule`. you can pass a custom name like:

Reviewer: ```suggestion
This makes an object available to JavaScript via the expression `NativeModules.SimpleHttpModule`. By default the JavaScript name will match the C# class name. If you don't want the name of the class to match the name in JavaScript, i.e. you want to access the module via the expression `NativeModules.CustomModule`. you can pass a custom name like:
```

```c#
  [ReactModule("CustomModule")]
  class SimpleHttpModule
  { 
    ...
  }
```

> Now we want to expose the method that performs the http request. It is recommended, and it is the default, to write these functions asynchroneously. Writing async code in C# is pretty straight-forward and intuitivie with the `async` and `await` keywords and the `Task<T>` types. 

Reviewer: ```suggestion
Now we want to expose the method that performs the HTTP request. It is recommended, and it is the default, to write these functions asynchronously. Writing async code in C# is pretty straight-forward and intuitive with the `async` and `await` keywords and the `Task<T>` types. 
```

> If you're not familiar with writing asynchronous C# code, see [Call asynchronous APIs in C# or Visual Basic](https://docs.microsoft.com/en-us/windows/uwp/threading-async/call-asynchronous-apis-in-csharp-or-visual-basic) and [Asynchronous programming](https://docs.microsoft.com/en-us/dotnet/csharp/async) that will teach you the concepts if you are not familiar yet.

The function signature for a typical web request, annotated with the `[ReactMethod]` attribute will look like:

```cs
    [ReactMethod]
    public async Task<string> GetHttpResponseAsync(string uri) {
      ...
    }
```

You are now free to fill in the logic like:

```cs
      // Create an HttpClient object
      var httpClient = new HttpClient();

      // Send the GET request asynchronously
      var httpResponseMessage = await httpClient.GetAsync(new Uri(uri));

      var statusCode = httpResponseMessage.StatusCode;
      var content = await httpResponseMessage.Content.ReadAsStringAsync();
      
      return content;
```

> Code code:

Reviewer: code code?

Author: ```suggestion
The code takes the following steps:
```

1. Creates a `HttpClient`.

> 2. Asynchronous calls the `GetAsync` method to make an HTTP request for the uri.

Reviewer: ```suggestion
2. Asynchronously calls the `GetAsync` method to make an HTTP request for the URI.
```

3. Parses the status code out of the returned `HttpResponseMessage` object.
4. Asynchronously parses the content out of the returned `HttpResponseMessage` object.
5. returns the content

This code only returns a string. You might want to return a more complex object that contains both the content and the status code.
For that you can simply declare a C# struct that will be marshalled to JavasScript like:

```cs
  internal struct Result {
    public int statusCode { get; set; }
    public string content { get; set; }
  }
```

It is recommended to follow JavaScript naming conventions here as of now there is no auto-mapping of names between the common style guides of C# and Js

To return the value you'll of course have to update the signature of the method from returning a `string` to the `Result`:

```cs
    public async Task<Result> GetHttpResponseAsync(string uri) {
```
as well as update the return statement from `return content;` to:

```cs
      return new Result()
      {
        statusCode = (int)statusCode,
        content = content,
      };
```

> But wait, we only discussed the success path, what happens if `GetHttpResponse` doesn't succeed? We don't handle any exceptions in this example, so if an exception is thrown, how do we marshal an error back to the JS? That is actually take care of for you by the framework, any exception in the task will be marshalled to the JavaScript side as a javascript exception.

Reviewer: ```suggestion
But wait, we've only discussed the success path, what happens if `GetHttpResponse` doesn't succeed? We don't handle any exceptions in this example, so if an exception is thrown, how do we marshal an error back to JavaScript? That is actually taken care of for you by the framework: any exception in the task will be marshalled to the JavaScript side as a JavaScript exception.
```
